Add nested hexagonal positions

GiudGiud opened the pull request:
refs #30240

GiudGiud committed:
Add nesting to hex grid positions
refs #30240

GiudGiud committed:
Add a regression test for hexagonal grid nesting
refs #30240

lewisgross1296 commented on modules/reactor/test/tests/positions/nested_hexagonal_grid_positions.i:
```diff
@@ -0,0 +1,72 @@
+    lattice_flat_to_flat = 4
+    pin_pitch = 0.5
+  []
+  [parent_grid]
+    type = HexagonalGridPositions
+    center = '0 0 0'
+    nr = 2
+    positions_pattern =           'hex_grid hex_grid;
+    hex_grid_2rings_exclusions  hex_grid_2rings_exclusions hex_grid;
+                    -1 hex_grid'
+    lattice_flat_to_flat = 12
+    pin_pitch = 4
+    outputs = 'out'
+  []
```
Would it be possible to have something similar to the `inputs` (maybe name it `positions_inputs`) parameter belonging to `PatterenedHexMeshGenerator` so that we could make a `HexagonalGridPositions` similarly by just using integers to represent the other `Positions` objects at each point in the `positions_pattern`. For example
```
  [full_core]
    type = HexagonalGridPositions
    positions_inputs = ' inner_assembly middle_assembly outer_assembly dummy'
                                      #  0            1             2            3
    positions_pattern = '
                    3 3 3 3 3 3 3;
                   3 3 2 2 2 2 3 3;
                  3 2 2 1 1 1 2 2 3;
                 3 2 1 1 1 1 1 1 2 3;
                3 2 1 1 0 0 0 1 1 2 3;
               3 2 1 1 0 0 0 0 1 1 2 3;
              3 3 2 1 0 0 0 0 0 1 2 3 3;
               3 2 1 1 0 0 0 0 1 1 2 3;
                3 2 1 1 0 0 0 1 1 2 3;
                 3 2 1 1 1 1 1 1 2 3;
                  3 2 2 1 1 1 2 2 3;
                   3 3 2 2 2 2 3 3;
                    3 3 3 3 3 3 3'
```
Perhaps `positions_inputs` is a required parameter if you specify `positions_pattern` and is not allowed if it is not supplied?

GiudGiud replied:
ok that works. 

I might do that instead:
'positions_pattern_indexes'

GiudGiud committed:
Add an index based input of the hex lattice
refs #30240

GiudGiud committed:
Adapt input syntax for nested inputs
Regold post fix of no-pattern input
refs #30240

GiudGiud commented on modules/reactor/src/positions/HexagonalGridPositions.C:
```diff
@@ -105,18 +148,76 @@ HexagonalGridPositions::initialize()
-  // Unroll pattern
-  std::vector<int> pattern_unrolled;
-  if (_pattern.size())
+  if (!isParamSetByUser("positions_pattern"))
```
no you can call initialize as many times a syou wnat

lindsayad replied:
then we can save some map lookups

GiudGiud replied:
we could catch the unrolling for sure (but cant cache the whole thing, the positions in the pattern could be dynamic)
if @lewisgross1296 sees the positions being slow i ll do it

lindsayad replied:
This comment was only in regard to constructing something like `_using_positions_pattern(isParamSetByUser("positions_pattern"))`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2795329/), step Docs: sync website on 1170191 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30231/site)

This comment will be updated on new commits.

GiudGiud committed:
Address Alex review

- use size_t for sizes over auto
- dont case size_t to int carelessly
- more !empty()

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2796328/), step Clang format on 923a9fd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30231/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30231/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a7b98aa272700a4a06703efd4da8a041454da026`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2796523/), step Generate coverage on d60c2eb wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>a7b98a</th>
      <th colspan=4>#30231 d60c2e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.28%</td>
      <td>85.28%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>109556</td>
      <td>109556</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18907</td>
      <td>18908</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30231/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30231/coverage/framework)


## Modules coverage

Inconsistent report tags were found between the head and base reports.
This can happen when reports are missing from either the head or the base.

Inconsistent tags:
  module-porous_flow-recover

### Reactor

<table>
  <thead>
    <tr>
      <th></th>
      <th>a7b98a</th>
      <th colspan=4>#30231 d60c2e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>93.67%</td>
      <td>93.59%</td>
      <td>-0.08%</td>
      <td>82.35%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7815</td>
      <td>7857</td>
      <td>+42</td>
      <td>56</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>528</td>
      <td>538</td>
      <td>+10</td>
      <td>12</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30231/coverage_diff/reactor)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30231/coverage/reactor)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30231/coverage/xfem)

</details>

## Warnings

- `reactor` new line coverage rate 82.35% is less than the suggested 90.0%



This comment will be updated on new commits.

miaoyinb reviewed (approved):
Looks good.
I think we should develop a `MeshMetaDataPositions` and corresponding meta data in MGs.

lewisgross1296 commented:
Was this supposed to close #30240 ? Or perhaps there's a bit more to do to fully close it
